mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessageVaultHook.java

```diff
@@ -0,0 +1,108 @@
+import com.google.common.collect.ImmutableList;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.GroupedFlux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class DeletedMessageVaultHook implements PreDeletionHook {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeletedMessageVaultHook.class);
+
+    private final MailboxSession session;
+    private final DeletedMessageVault deletedMessageVault;
+    private final DeletedMessageConverter deletedMessageConverter;
+    private final MailboxSessionMapperFactory mapperFactory;
+
+    DeletedMessageVaultHook(SessionProvider sessionProvider,
+                            DeletedMessageVault deletedMessageVault,
+                            DeletedMessageConverter deletedMessageConverter,
+                            MailboxSessionMapperFactory mapperFactory) {
+        this.session = sessionProvider.createSystemSession(getClass().getName());
+        this.deletedMessageVault = deletedMessageVault;
+        this.deletedMessageConverter = deletedMessageConverter;
+        this.mapperFactory = mapperFactory;
+    }
+
+    @Override
+    public Publisher<Void> notifyDelete(DeleteOperation deleteOperation) {
+        Preconditions.checkNotNull(deleteOperation);
+
+        return groupMetadataByOwnerAndMessageId(deleteOperation)
+            .flatMap(Throwing.<DeletedMessageMetadata, Mono<Void>>function(entry -> moveIntoTheVault(entry)).sneakyThrow())
+            .then();
+    }
+
+    private Mono<Void> moveIntoTheVault(DeletedMessageMetadata deletedMessageMetadata) throws MailboxException, IOException {
+        Optional<MailboxMessage> mailboxMessage = mapperFactory.getMessageIdMapper(session).find(ImmutableList.of(deletedMessageMetadata.getMessageId()), MessageMapper.FetchType.Full).stream().findFirst();
+
+        if (mailboxMessage.isPresent()) {
+            DeletedMessage deletedMessage = deletedMessageConverter.convert(deletedMessageMetadata, mailboxMessage.get());
+            return Mono.from(deletedMessageVault.append(deletedMessage.getOwner(), deletedMessage, mailboxMessage.get().getFullContent()));
+        }
+        LOGGER.warn("Can not find mailbox message with msgId {}", deletedMessageMetadata.getMessageId());
+
+        return Mono.empty();
+    }
+
+    private Flux<DeletedMessageMetadata> groupMetadataByOwnerAndMessageId(DeleteOperation deleteOperation) {
+        return Flux.fromIterable(deleteOperation.getDeletionMetadataList())
+            .publishOn(Schedulers.elastic())
+            .map(Throwing.function(this::toMetadataWithOwner))
+            .groupBy(DeletedMessageMetadata::groupingKey)
+            .flatMap(this::toDeletedMessageMetadata);
+    }
+
+    private Mono<DeletedMessageMetadata> toDeletedMessageMetadata(GroupedFlux<Pair<MessageId, User>, DeletedMessageMetadata> groupedFlux) {
+        return groupedFlux.reduce(DeletedMessageMetadata::combineWith);
```

Comment:
What does it mean to reduce several DeletedMessageMetadata to a single one?

Reply:
idem

Reply:
Well I guess the misconception here is that DeletedMessageMetadata is a stringly typed Tuple<3> representing the mailbox context of a deleted message

I propose :
  - Renaming it to `DeletedMessageMailboxContext`
  - Moving that class as an `inner class` in the Hook
  - putting the `combineWith` & `groupingKey` method private to ensure only the hook can access them - (as it do not make sense in another context)

Would that be satsifying for you @mbaechler ?

Reply:
> Would that be satsifying for you @mbaechler ?

I feel that you got my point but I don't understand your solution enough to give you a yes. AFAIU it looks ok.